File: problemSolving/max-sum-subarray/max-sum-subarray.cpp
#include <iostream>
#include <vector>


std::vector<int> findMaxSumSubarray(std::vector<int> &x, int indexStart, int indexEnd);
std::vector<int> findMaxCrossSubarray(std::vector<int> &x, int indexStart, int indexMiddle, int indexEnd);
int sumVector(std::vector<int> &x);
void printVector(std::vector<int> &x);


int main(){
    
    std::vector<int> x = {13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7};
    
    std::vector<int> subX = findMaxSumSubarray(x, 0, x.size() - 1);
    
    printVector(subX);
    std::cout << sumVector(subX) << std::endl;
    
    return 0;
}


std::vector<int> findMaxSumSubarray(std::vector<int> &x, int indexStart, int indexEnd){
    
    if(indexStart < indexEnd){
        
        int indexMiddle = indexStart + (indexEnd - indexStart) / 2;
    
        std::vector<int> left = findMaxSumSubarray(x, indexStart, indexMiddle);
        std::vector<int> right = findMaxSumSubarray(x, indexMiddle + 1, indexEnd);
        std::vector<int> cross = findMaxCrossSubarray(x, indexStart, indexMiddle, indexEnd);
        
        int sumLeft = sumVector(left);
        int sumRight = sumVector(right);
        int sumCross = sumVector(cross);
        
        if(sumRight > sumLeft && sumRight > sumCross)
            return right;
            
        else if (sumLeft > sumRight && sumLeft > sumCross)
            return left;
        
        else
            return cross;
    }
    
    std::vector<int> empty = {};
    
    return empty;
}


std::vector<int> findMaxCrossSubarray(std::vector<int> &x, int indexStart, int indexMiddle, int indexEnd){

    // find max left subarray that ends at indexMiddle
    int indexMaxLeft = indexMiddle;
    int maxSumLeft = 0;
    int currentSum = 0;
    for(int i = indexMaxLeft; i >= 0; i--){
        
        currentSum += x[i];
        
        if(currentSum > maxSumLeft){
            maxSumLeft = currentSum;
            indexMaxLeft = i;
        }
    }

    // find max right subarray that starts at indexMiddle + 1
    int indexMaxRight = indexMiddle + 1;
    int maxSumRight = 0;
    currentSum = 0;
    for(int i = indexMaxRight; i <= indexEnd; i++){
        
        currentSum += x[i];
        
        if(currentSum > maxSumRight){
            maxSumRight = currentSum;
            indexMaxRight = i;
        }
    }

    // get maximum cross subarray
    std::vector<int> maxCrossSubarray;
    for(int i = indexMaxLeft; i <= indexMaxRight; i++)
        maxCrossSubarray.push_back(x[i]);

    return maxCrossSubarray;
}



int sumVector(std::vector<int> &x){
    
    int n = x.size();
    int sum = 0;
    
    for(int i = 0; i < n; i++)
        sum += x[i];
    
    return sum;
}


void printVector(std::vector<int> &x){
    
    int n = x.size();
    
    for(int i = 0; i < n; i++)
        std::cout << x[i] << " ";
        
     std::cout << std::endl;
}

File: max-sum-subarray/max-sum-subarray.py
"""

Given an array, find the contiguous subarray s.t. sum of elements is maximal

"""


def find_max_sum_subarray_brute_force(x):

    max_sum, index_max_start, index_max_end = -float('inf'), None, None

    for index_start in range(len(x)):
        current_sum = 0

        for index_end in range(index_start, len(x)):
            current_sum += x[index_end]

            if current_sum > max_sum:
                max_sum, index_max_start, index_max_end = current_sum, index_start, index_end

    return x[index_max_start:(index_max_end + 1)]


def find_max_sum_subarray_divide_conquer(x, index_start, index_end):
    if index_start < index_end:
        index_middle = index_start + (index_end - index_start) / 2

        left = find_max_sum_subarray_divide_conquer(x, index_start, index_middle)
        right = find_max_sum_subarray_divide_conquer(x, index_middle + 1, index_end)
        cross = _find_max_cross_subarray(x, index_start, index_middle, index_end)

        sum_left = sum(left)
        sum_right = sum(right)
        sum_cross = sum(cross)

        if sum_left >= sum_right and sum_left >= sum_cross:
            return left
        elif sum_right >= sum_left and sum_right >= sum_cross:
            return right
        else:
            return cross

    else:
        single_value = x[index_start]
        return [single_value]


def _find_max_cross_subarray(x, index_start, index_middle, index_end):
    #  Find max subarray [?, index_middle]
    max_left, index_max_left = -float('inf'), index_middle

    sum_left = 0
    for index_left in range(index_middle, index_start - 1, -1):
        sum_left += x[index_left]

        if sum_left > max_left:
            max_left, index_max_left = sum_left, index_left

    # Find max subarray [index_middle + 1, ?]
    max_right, index_max_right = -float('inf'), index_middle + 1

    sum_right = 0
    for index_right in range(index_middle + 1, index_end + 1):
        sum_right += x[index_right]

        if sum_right > max_right:
            max_right, index_max_right = sum_right, index_right

    return x[index_max_left:(index_max_right + 1)]


def find_max_sum_subarray_dynamic_programming(x):
    max_sum, index_max_left, index_max_right = -float('inf'), None, None

    current_sum, index_left = 0, 0
    for index_right in range(len(x)):

        current_sum += x[index_right]

        #  keep track of maximal subarray ENDING at index_right
        if current_sum <= x[index_right]:
            current_sum, index_left = x[index_right], index_right

        if current_sum > max_sum:
            max_sum, index_max_left, index_max_right = current_sum, index_left, index_right

    return x[index_max_left:(index_max_right + 1)]


if __name__ == '__main__':
    x = [13, -3, -25, 20, -3, -16, -23, 18, 20, -7, 12, -5, -22, 15, -4, 7]
    print 'Array: ' + str(x)
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_brute_force(x))
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_divide_conquer(x, 0, len(x) - 1))
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_dynamic_programming(x))

    x = [-13, -3, -25, -20, -3, -16, -23, -18, -20, -7, -12, -5, -22, -15, -4, -7]
    print 'Array: ' + str(x)
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_brute_force(x))
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_divide_conquer(x, 0, len(x) - 1))
    print 'Max Sum Subarray: ' + str(find_max_sum_subarray_dynamic_programming(x))


